Scala3 Documentation (#3)
scalacenter/docs.scala-lang: _overviews/scala3-book/domain-modeling-tools.md

---
title: Tools
type: section
description: This chapter provides an introduction to the available domain modeling tools in Scala 3, including classes, traits, enums, and more.
num: 20
previous-page: domain-modeling-intro
next-page: domain-modeling-oop
---

> Scala 3 provides the following tools to help us model the world around us:

Reviewer: ```suggestion
Scala 3 provides us with the following tools to model the world around us:
```

- Class
- Companion object
- Trait
- Abstract class
- Enum
- Case class
- Case object

The following sections introduce each of these modeling tools.

## Classes

> As with other languages, a Scala *class* is a template for the creation of object instances. Here are some examples of classes that have constructor parameters, but no fields or methods:

Reviewer: IIRC specifying `var` means that you are creating fields in the class. To only have constructor parameters, but no fields or methods, you should drop the `var`s

Author: @alvinj if you change the code to address @vincenzobaz comment, then also the text following needs adjustment since it talks about var fields. In fact,
```scala
class A(var x: T)
```
creates a field `x` and a constructor expecting a `T` and assigning to `x`.

Reviewer 2: This was re-worded, so I’m resolving this conversation.

```scala
class Person(var name: String, var vocation: String)
class Book(var title: String, var author: String, var year: Int)
class Movie(var name: String, var director: String, var year: Int)
```

All of those parameters are defined as `var` fields, which means they are mutable: you can read them, and also modify them. If you want them to be immutable — read only — create them as `val` fields instead.

Prior to Scala 3, you used the `new` keyword to create a new instance of a class:

```scala
val p = new Person("Robert Allen Zimmerman", "Harmonica Player")
        ---
```

However, this isn’t required in Scala 3:

```scala

> val p = new Person("Robert Allen Zimmerman", "Harmonica Player")

Reviewer: ```suggestion
val p = Person("Robert Allen Zimmerman", "Harmonica Player")
```

```

Once you have an instance of a class, you can access its fields, which in this example are all constructor parameters:

```scala
p.name       // "Robert Allen Zimmerman"
p.vocation   // "Harmonica Player"
```

As mentioned, all of these parameters were created as `var` fields, so you can also mutate them:

```scala
p.name = "Bob Dylan"
p.vocation = "Musician"
```

### Fields and methods

In addition to constructor parameters, classes can also have fields and methods. They are defined in the body of the class, which also serves as the class constructor:

```scala
class Person(var firstName: String, var lastName: String):
  
  println("the constructor begins")
  val fullName = firstName + " " + lastName

  // a class method
  def printFullName: Unit =
    // access the `fullName` field, which is created above
    println(fullName)
  
  printFullName
  println("this ends the constructor")
```

> The REPL shows how to create a new `Person` instance with this class:

Reviewer: ```suggestion
This REPL extract shows how to create a new `Person` instance with this class:
```

Author: maybe you mean "excerpt" instead of "extract"?

````
scala> val john = Person("John", "Doe")
the constructor begins
John Doe
this ends the constructor
val john: Person = Person@55d8f6bb

scala> john.printFullName
John Doe
````

> 

Reviewer: Maybe mention that you can define additional constructors with `def this...`

Author: Not sure people _should_ know that ;) No, you are right -- sounds like a good idea.

Classes can also extend traits and abstract classes. Examples of this are shown in the Trait and Abstract Class sections.

### Default parameter values

As a quick look at a few other features, class constructor parameters can also have default values:

```scala
class Socket(
  val timeout: Int = 5_000,
  val linger: Int = 5_000
):
    override def toString = s"timeout: $timeout, linger: $linger"
```

A great thing about this feature is that it lets consumers of your code create classes in a variety of different ways, as though the class had alternate constructors:

```scala
val s1 = Socket()                 // timeout: 5000, linger: 5000
val s2 = Socket(2_500)            // timeout: 2500, linger: 5000
val s3 = Socket(10_000, 10_000)   // timeout: 10000, linger: 10000

> ```

Reviewer: ```suggestion
val s4 = Socket(linger=10_000)
```
```

When creating a new instance of a class, you can also use named parameters. This is particularly helpful when many of the parameters have the same data type:

```scala
val s = Socket(
  timeout = 10_000,
  linger = 10_000
)
```

## Objects

> An object is a class that has exactly one instance. It’s created lazily when it is referenced, like a `lazy val`. If you’re familiar with Java, methods and fields in an object are accessed like `static` members in Java.

Reviewer: ```suggestion
An object is a class that has exactly one instance. It’s created lazily when it is referenced, like a `lazy val`. If you’re familiar with Java, methods and fields in an object behave like `static` members in Java, Javascript (ES6) or `@staticmethod` in Python.
```

Author: ```suggestion
An object is a class that has exactly one instance. It’s initialized lazily when its members are referenced, similar to a `lazy val`. Objects in Scala allow grouping methods and fields under one namespace, similar to how can use `static` members on a class in Java, Javascript (ES6) or `@staticmethod` in Python
```

- You want to create a base class that takes constructor arguments
- The code will be called from Java code

### A base class that takes constructor arguments

Prior to Scala 3, when a base class needed to take constructor arguments, you’d declare it as an `abstract class`:

```scala
abstract class Pet(var name: String):
  def greeting: String
  def age: Int
  override def toString = s"I say $greeting, and I’m $age"

class Dog(name: String, var age: Int) extends Pet(name):
  val greeting = "Woof"

val d = Dog("Fido", 1)
```

However, with Scala 3, traits can now have parameters, so you can now use a trait in the same situation:

```scala
trait Pet(var name: String):
  def greeting: String
  def age: Int
  override def toString = s"I say $greeting, and I’m $age"

class Dog(name: String, var age: Int) extends Pet(name):
  val greeting = "Woof"

val d = Dog("Fido", 1)
```

### When a Scala abstract class is called from Java

{% comment %}
TODO: I need to add content here.
{% endcomment %}

## Enums

> An enumeration is used to define a type that consists of a set of named values. Basic enumerations are used to define sets of constants, like the months in a year, the days in a week, directions like north/south/east/west, and more.

Reviewer: ```suggestion
An enumeration is used to define a type that consists of a finite set of named values. Basic enumerations are used to define sets of constants, like the months in a year, the days in a week, directions like north/south/east/west, and more.
```

As an example, these enumerations define sets of attributes related to pizzas:

```scala
enum CrustSize:
  case Small, Medium, Large

enum CrustType:
  case Thin, Thick, Regular

enum Topping:
  case Cheese, Pepperoni, BlackOlives, GreenOlives, Onions
```

To use them in other code, first import them, and then use them:

```scala
import CrustSize._
val currentCrustSize = Small
```

They can be used anywhere classes can be used, including if/then expressions, match expressions, and more:

```scala
// if/then
if (currentCrustSize == Large)
  println("You get a prize!")

// match
currentCrustSize match
  case Small => println("small")
  case Medium => println("medium")
  case Large => println("large")
```

In the modeling sections that follow, you’ll also see that they can be used as constructor parameters, and to model ADTs and GADTs.

### More enum features

Enumerations have more features. You have the basic enum just shown:

```scala
enum Color:
  case Red, Green, Blue
```

They can also be parameterized:

```scala
enum Color(val rgb: Int):
  case Red   extends Color(0xFF0000)
  case Green extends Color(0x00FF00)
  case Blue  extends Color(0x0000FF)
```

> And they can also have user-defined members:

Reviewer: ```suggestion
And they can also have members:
```

```scala
enum Color extends Enum[Color] { case Red, Green, Blue }
```

The type parameter comes from the Java `enum` definition, and should be the same as the type of the enum. There’s no need to provide constructor arguments (as defined in the Java API docs) to *java.lang.Enum* when extending it — the compiler generates them automatically.

After defining `Color` like that, you can use it like you would a Java enum:

````
scala> Color.Red.compareTo(Color.Green)
val res0: Int = -1
````

## Case classes

A `case class` has all of the functionality of a `class`, and more. When the compiler sees the `case` keyword in front of a `class`, it generates code for you, with the following benefits:

{% comment %}
TODO: What to say about `apply` methods in the following bullet points? They’re less significant in Scala 3.
{% endcomment %}

* Case class constructor parameters are public `val` fields by default, so accessor methods are generated for each parameter
* An `apply` method is created in the companion object of the class, so you don’t need to use the `new` keyword to create a new instance of the class

> * An `unapply` method is generated, which lets you use case classes in more ways in `match` expressions

Reviewer: IIRC the requirements for pattern matching changed in scala 3 as well

Author: Yes, there is this document describing "Optionless" pattern matching: http://dotty.epfl.ch/docs/reference/changed-features/pattern-matching.html

Reviewer 2: In trying to resolve this, is there something specific we should add here? Or should we create another case class discussion elsewhere?

Author: I think we could discuss it here, but I would defer this to a revisiin, after we release the docs

case Teacher(name, whatTheyTeach) =>
```

Those patterns work because `Student` and `Teacher` are defined as case classes that have `unapply` methods whose type signature conforms to a certain standard. Technically, the specific type of pattern matching shown in these examples is known as a *constructor pattern*.

>The Scala standard is that an `unapply` method returns the case class constructor fields in a tuple that’s wrapped in an `Option`. The “tuple” part of the solution was shown in the previous lesson.

To show how that code works, create an instance of `Student` and `Teacher`:

```scala
val s = Student("Al", 1)
val t = Teacher("Bob Donnan", "Mathematics")
```

Next, this is what the output looks like in the REPL when you call `getPrintableString` with those two instances:

```scala
scala> getPrintableString(s)
res0: String = Al is a student in Year 1.

scala> getPrintableString(t)
res1: String = Bob Donnan teaches Mathematics.
```

>All of this content on `unapply` methods and extractors is a little advanced for an introductory book like this, but because case classes are an important FP topic, it seems better to cover them, rather than skipping over them.
{% endcomment %}

> 

Reviewer: You could show how to add pattern matching to any type by implementing `unapply`

Reviewer 2: I added an `unapply` example. This text is actually inside a comment block. Would you (all) prefer that I remove the comments and make it part of the page?

## Case objects

Case objects are to objects what case classes are to classes: they provide a number of automatically-generated methods to make them more powerful. They’re particularly useful whenever you need a singleton object that needs a little extra functionality, such as being used with pattern matching in `match` expressions.

Case objects are useful when you need to pass immutable messages around. For instance, if you’re working on a music player project, you’ll create a set of commands or messages like this:

```scala
sealed trait Message
case object PlaySong(name: String) extends Message
case object IncreaseVolume(amount: Int) extends Message
case object DecreaseVolume(amount: Int) extends Message
case object StopPlaying extends Message
```

Then in other parts of your code you can write methods like this, which use pattern matching to handle the incoming message:

```scala
def handleMessages(msg: Message) = message match
  case PlaySong(name)      => playSong(name)
  case IncreaseVolume(amt) => changeVolume(amt)
  case DecreaseVolume(amt) => changeVolume(-amt)
  case StopPlaying         => stopPlayingMusic
```

> 

Reviewer: I would mention that
```scala
val ps: PlaySong = PlaySong("Yellow submarine")
val sp: StopPlaying.type = StopPlaying
```
